src/dryrun.ts

```diff
@@ -167,3 +171,293 @@ export async function createDryrun({
+  logs: string[] | undefined;
+  appCallTrace: DryrunTrace | undefined;
+  logicSigTrace: DryrunTrace | undefined;
+}
+
+interface StackPrinterConfig {
+  maxWidth: number | undefined;
+  topOfStackFirst: boolean | undefined;
+}
+
+function truncate(str: string, maxWidth: number): string {
+  if (str.length > maxWidth && maxWidth > 0) {
+    return `${str.slice(0, maxWidth)}...`;
+  }
+  return str;
+}
+
+function scratchToString(
+  prevScratch: TealValue[],
+  currScratch: TealValue[]
+): string {
+  if (currScratch.length === 0) return '';
+
+  let newScratchIdx = null;
+  for (let idx = 0; idx < currScratch.length; idx++) {
+    if (idx > prevScratch.length) {
+      newScratchIdx = idx;
+      continue;
+    }
+
+    if (JSON.stringify(prevScratch[idx]) !== JSON.stringify(currScratch[idx])) {
+      newScratchIdx = idx;
+    }
+  }
+
+  if (newScratchIdx == null) return '';
+
+  const newScratch = currScratch[newScratchIdx];
+  if (newScratch.bytes.length > 0) {
+    return `${newScratchIdx} = 0x${Buffer.from(
+      newScratch.bytes,
+      'base64'
+    ).toString('hex')}`;
+  }
+  return `${newScratchIdx} = ${newScratch.uint.toString()}`;
+}
+
+function stackToString(stack: DryrunStackValue[], reverse: boolean): string {
+  const svs = reverse ? stack.reverse() : stack;
+  return `[${svs
+    .map((sv: DryrunStackValue) => {
+      switch (sv.type) {
+        case 1:
+          return `0x${Buffer.from(sv.bytes, 'base64').toString('hex')}`;
+        case 2:
+          return `${sv.uint.toString()}`;
+        default:
+          return '';
+      }
+    })
+    .join(', ')}]`;
+}
+
+class DryrunTransactionResult {
+  disassembly: string[] = [];
+  appCallMessages: string[] | undefined = [];
+  localDeltas: AccountStateDelta[] | undefined = [];
+  globalDelta: EvalDeltaKeyValue[] | undefined = [];
+  cost: number | undefined = 0;
+  logicSigMessages: string[] | undefined = [];
+  logicSigDisassemly: string[] | undefined = [];
+  logs: string[] | undefined = [];
+
+  appCallTrace: DryrunTrace | undefined = undefined;
+  logicSigTrace: DryrunTrace | undefined = undefined;
+
+  required = ['disassembly'];
+  optionals = [
+    'app-call-messages',
+    'local-deltas',
+    'global-delta',
+    'cost',
+    'logic-sig-messages',
+    'logic-sig-disassembly',
+    'logs',
+  ];
+
+  traces = ['app-call-trace', 'logic-sig-trace'];
+
+  constructor(dtr: DryrunTransactionResultResponse) {
+    this.disassembly = dtr.disassembly;
+    this.appCallMessages = dtr['app-call-messages'];
+    this.localDeltas = dtr['local-deltas'];
+    this.globalDelta = dtr['global-delta'];
+    this.cost = dtr.cost;
+    this.logicSigMessages = dtr['logic-sig-messages'];
+    this.logicSigDisassemly = dtr['logic-sig-disassembly'];
+    this.logs = dtr.logs;
+    this.appCallTrace = new DryrunTrace(dtr['app-call-trace']);
+    this.logicSigTrace = new DryrunTrace(dtr['logic-sig-trace']);
+  }
+
+  appCallRejected(): boolean {
```

Comment:
Are these functions used somewhere?

Reply:
Only convenience methods for the user to do something like iterate over txns and print trace for only those that failed